lib/psql.js

```diff
@@ -24,31 +25,123 @@ function env (db) {
   })
 }
 
+function tunnelConfig (db) {
+  const localHost = '127.0.0.1'
+  const localPort = Math.floor(Math.random() * (65535 - 49152) + 49152)
+  return {
+    username: 'bastion',
+    host: db.bastionHost,
+    privateKey: db.bastionKey,
+    dstHost: db.host,
+    dstPort: db.port,
+    localHost: localHost,
+    localPort: localPort
+  }
+}
+
+function buildTunnelPromise (db, dbTunnelConfig, timeout) {
+  return new Promise((resolve, reject) => {
+    // if necessary to tunnel, setup a tunnel
+    // see also https://github.com/heroku/heroku/blob/master/lib/heroku/helpers/heroku_postgresql.rb#L53-L80
+    if (db.bastionKey) {
+      tunnel(dbTunnelConfig, (err, tnl) => {
+        if (err) {
+          debug(err)
+          cli.error(`Unable to establish a secure tunnel to your database.`)
+          process.exit(1)
+        }
+        debug('Tunnel created')
+        resolve(tnl)
+      })
+    } else {
+      resolve()
+    }
+    setTimeout(() => reject('Establishing a secure tunnel timed out'), timeout)
+  }).catch((error) => {
+    cli.error(error)
+  })
+}
+
+function buildPsqlPromise (query, dbEnv, timeout) {
+  const stripEOF = require('strip-eof')
+  const {spawn} = require('child_process')
+  return new Promise((resolve, reject) => {
+    let psql = spawn('psql', ['-c', query], {env: dbEnv, encoding: 'utf8'})
+    psql.stdout.on('data', (data) => {
```

Comment:
I worry this might not work if the output is large. Maybe we could test it with the `pg:mandelbrot` repeated a few times?